matrix_sdk/src/http_client.rs

```diff
@@ -106,6 +108,25 @@ pub(crate) struct HttpClient {
 use crate::OutgoingRequestAppserviceExt;
 
 impl HttpClient {
+    pub(crate) fn new(
+        inner: Arc<dyn HttpSend>,
+        homeserver: Arc<Url>,
+        session: Arc<RwLock<Option<Session>>>,
+        request_config: RequestConfig,
+    ) -> Self {
+        HttpClient { inner, homeserver, session, request_config }
+    }
+
+    pub(crate) async fn new_with_well_known(
```

Comment:
This doesn't really do what's described in the spec, as far as I can tell it uses well-known if it can find it, otherwise it continues to use the domain from the `UserId`.

It should only continue to use the domain from the `UserId` if the well-known request returns a 404 error, otherwise it should throw an error, take a look at all the different error cases listed [here](https://spec.matrix.org/unstable/client-server-api/#well-known-uri).

Also please take a look how it should validate that the now modified `Url` is a valid homeserver using the `/versions` endpoint over [here](https://spec.matrix.org/unstable/client-server-api/#well-known-uri), step 5, substep 2.

Reply:
Ohh, there is a formal spec, how cool!

After reading the spec I don't think it should ever use the domain from the `UserId` – at best it can fall back to "other discovery mechanisms", but this is an empty set. This is a bit of a moot point too, since at least synapse automatically provides a `.well-known...` endpoint with a self-reference.

Validation added.

I also got inspired by the spec to improve the design a bit in general. All the work actually happens in `Client` now and there is a method to change the homeserver url in there – private for now, but if I understand correctly it might be useful for the other well-known case you mentioned (the one  #219 originally referred to?).